File: dotnet/src/Microsoft.Agents.Protocol.Runtime/Tools/ToolExecutor.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Reflection;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;

namespace Microsoft.Agents.Protocol.Runtime.Tools;

/// <summary>
/// Executes tools/functions with JSON arguments.
/// Shared utility across Client and Hosting SDKs for consistent tool execution.
/// </summary>
/// <remarks>
/// This class provides centralized tool execution logic, eliminating duplication
/// between Client SDK's ToolDefinition.ExecuteAsync and Hosting SDK's function
/// execution. Handles both synchronous and asynchronous functions, parameter
/// binding, and error handling.
/// </remarks>
public static class ToolExecutor
{
    /// <summary>
    /// Executes a tool handler with JSON arguments.
    /// </summary>
    /// <param name="handler">The delegate to execute</param>
    /// <param name="argumentsJson">JSON-encoded arguments</param>
    /// <param name="cancellationToken">Cancellation token</param>
    /// <returns>Tool execution result</returns>
    /// <exception cref="ArgumentNullException">If handler or argumentsJson is null</exception>
    /// <exception cref="ArgumentException">If required parameter is missing</exception>
    /// <exception cref="JsonException">If JSON is invalid</exception>
    /// <example>
    /// <code>
    /// Func&lt;string, int, string&gt; handler = (name, age) =&gt; $"{name} is {age}";
    /// string json = "{\"name\": \"Alice\", \"age\": 30}";
    /// var result = await ToolExecutor.ExecuteAsync(handler, json);
    /// // result == "Alice is 30"
    /// </code>
    /// </example>
    public static async Task<object> ExecuteAsync(
        Delegate handler,
        string argumentsJson,
        CancellationToken cancellationToken = default)
    {
        if (handler == null)
            throw new ArgumentNullException(nameof(handler));

        if (string.IsNullOrEmpty(argumentsJson))
            throw new ArgumentNullException(nameof(argumentsJson));

        var method = handler.Method;
        var parameters = method.GetParameters();

        // Parse JSON arguments
        JsonDocument jsonDoc;
        try
        {
            jsonDoc = JsonDocument.Parse(argumentsJson);
        }
        catch (JsonException ex)
        {
            throw new ArgumentException($"Invalid JSON arguments: {ex.Message}", nameof(argumentsJson), ex);
        }

        using (jsonDoc)
        {
            var args = new object?[parameters.Length];

            // Bind parameters from JSON
            for (int i = 0; i < parameters.Length; i++)
            {
                var param = parameters[i];
                if (param.Name == null)
                    continue;

                if (jsonDoc.RootElement.TryGetProperty(param.Name, out var value))
                {
                    try
                    {
                        args[i] = JsonSerializer.Deserialize(value.GetRawText(), param.ParameterType);
                    }
                    catch (JsonException ex)
                    {
                        throw new ArgumentException(
                            $"Failed to deserialize parameter '{param.Name}' to type {param.ParameterType.Name}: {ex.Message}",
                            nameof(argumentsJson),
                            ex);
                    }
                }
                else if (param.IsOptional || param.HasDefaultValue)
                {
                    args[i] = param.DefaultValue;
                }
                else
                {
                    throw new ArgumentException(
                        $"Missing required parameter: {param.Name}",
                        nameof(argumentsJson));
                }
            }

            // Invoke the handler
            object? result;
            try
            {
                result = handler.DynamicInvoke(args);
            }
            catch (TargetInvocationException ex)
            {
                // Unwrap the inner exception to preserve original exception type
                if (ex.InnerException != null)
                    throw ex.InnerException;
                throw;
            }

            // Handle async results
            if (result is Task task)
            {
                await task.WaitAsync(cancellationToken);

                // Extract result from Task<T>
                var resultProperty = task.GetType().GetProperty("Result");
                return resultProperty?.GetValue(task) ?? string.Empty;
            }

            return result ?? string.Empty;
        }
    }

    /// <summary>
    /// Executes a tool handler with JSON arguments and returns a string result.
    /// </summary>
    /// <param name="handler">The delegate to execute</param>
    /// <param name="argumentsJson">JSON-encoded arguments</param>
    /// <param name="cancellationToken">Cancellation token</param>
    /// <returns>String representation of the result</returns>
    /// <example>
    /// <code>
    /// Func&lt;int, int, int&gt; handler = (a, b) =&gt; a + b;
    /// string json = "{\"a\": 5, \"b\": 3}";
    /// var result = await ToolExecutor.ExecuteAsStringAsync(handler, json);
    /// // result == "8"
    /// </code>
    /// </example>
    public static async Task<string> ExecuteAsStringAsync(
        Delegate handler,
        string argumentsJson,
        CancellationToken cancellationToken = default)
    {
        var result = await ExecuteAsync(handler, argumentsJson, cancellationToken);

        return result switch
        {
            string s => s,
            null => string.Empty,
            _ => result.ToString() ?? string.Empty
        };
    }
}

File: typescript/packages/agents-protocol-runtime/src/tools/ToolExecutor.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import type { JSONSchema } from '@microsoft/agents-protocol-abstractions';

/**
 * Executes tools/functions with JSON arguments.
 * Shared utility across Client and Hosting SDKs for consistent tool execution.
 *
 * @example
 * ```typescript
 * const handler = async (params: { name: string; age: number }) => {
 *   return `${params.name} is ${params.age}`;
 * };
 *
 * const schema = {
 *   type: 'object',
 *   properties: {
 *     name: { type: 'string' },
 *     age: { type: 'number' }
 *   },
 *   required: ['name', 'age']
 * };
 *
 * const result = await ToolExecutor.execute(
 *   handler,
 *   schema,
 *   '{"name": "Alice", "age": 30}'
 * );
 * // result === "Alice is 30"
 * ```
 */
export class ToolExecutor {
  /**
   * Executes a tool with JSON arguments.
   *
   * @param handler - The function to execute
   * @param schema - JSON schema for validation
   * @param argumentsJson - JSON-encoded arguments
   * @returns Tool execution result as string
   * @throws Error if validation fails or execution errors
   *
   * @example
   * ```typescript
   * const handler = ({ a, b }: { a: number; b: number }) => a + b;
   * const schema = {
   *   type: 'object',
   *   properties: {
   *     a: { type: 'number' },
   *     b: { type: 'number' }
   *   },
   *   required: ['a', 'b']
   * };
   *
   * const result = await ToolExecutor.execute(
   *   handler,
   *   schema,
   *   '{"a": 5, "b": 3}'
   * );
   * // result === "8"
   * ```
   */
  static async execute<TParams = Record<string, unknown>>(
    handler: (params: TParams) => string | Promise<string>,
    schema: JSONSchema,
    argumentsJson: string
  ): Promise<string> {
    // Parse JSON arguments
    let args: unknown;
    try {
      args = JSON.parse(argumentsJson);
    } catch (error) {
      throw new Error(
        `Invalid JSON arguments: ${(error as Error).message}`
      );
    }

    // Validate arguments against schema
    ToolExecutor.validateArguments(args, schema);

    // Execute handler
    try {
      const result = await handler(args as TParams);
      return result;
    } catch (error) {
      // Preserve original error
      throw error;
    }
  }

  /**
   * Validates arguments against a JSON schema.
   *
   * @param args - The arguments to validate
   * @param schema - The schema to validate against
   * @throws Error if validation fails
   *
   * @example
   * ```typescript
   * const schema = {
   *   type: 'object',
   *   properties: { name: { type: 'string' } },
   *   required: ['name']
   * };
   *
   * ToolExecutor.validateArguments({ name: 'Alice' }, schema); // OK
   * ToolExecutor.validateArguments({}, schema); // Throws (missing 'name')
   * ```
   */
  static validateArguments(args: unknown, schema: JSONSchema): void {
    // Type validation
    if (schema.type === 'object' && typeof args !== 'object') {
      throw new Error(`Expected object, got ${typeof args}`);
    }

    if (schema.type === 'array' && !Array.isArray(args)) {
      throw new Error(`Expected array, got ${typeof args}`);
    }

    if (schema.type === 'string' && typeof args !== 'string') {
      throw new Error(`Expected string, got ${typeof args}`);
    }

    if (schema.type === 'number' && typeof args !== 'number') {
      throw new Error(`Expected number, got ${typeof args}`);
    }

    if (schema.type === 'boolean' && typeof args !== 'boolean') {
      throw new Error(`Expected boolean, got ${typeof args}`);
    }

    // Required properties validation
    if (schema.type === 'object' && schema.required) {
      const obj = args as Record<string, unknown>;
      for (const requiredProp of schema.required) {
        if (!(requiredProp in obj)) {
          throw new Error(`Missing required parameter: ${requiredProp}`);
        }
      }
    }

    // Property type validation
    if (schema.type === 'object' && schema.properties) {
      const obj = args as Record<string, unknown>;
      for (const [key, propSchema] of Object.entries(schema.properties)) {
        if (key in obj) {
          const value = obj[key];
          ToolExecutor.validateValue(value, propSchema, key);
        }
      }
    }

    // Array items validation
    if (schema.type === 'array' && schema.items) {
      const arr = args as unknown[];
      for (let i = 0; i < arr.length; i++) {
        ToolExecutor.validateValue(arr[i], schema.items, `[${i}]`);
      }
    }

    // Enum validation
    if (schema.enum && !schema.enum.includes(args)) {
      throw new Error(
        `Value must be one of: ${schema.enum.join(', ')}`
      );
    }

    // Number range validation
    if (schema.type === 'number' || schema.type === 'integer') {
      const num = args as number;
      if (schema.minimum !== undefined && num < schema.minimum) {
        throw new Error(`Value ${num} is less than minimum ${schema.minimum}`);
      }
      if (schema.maximum !== undefined && num > schema.maximum) {
        throw new Error(`Value ${num} is greater than maximum ${schema.maximum}`);
      }
    }

    // String length validation
    if (schema.type === 'string') {
      const str = args as string;
      if (schema.minLength !== undefined && str.length < schema.minLength) {
        throw new Error(
          `String length ${str.length} is less than minimum ${schema.minLength}`
        );
      }
      if (schema.maxLength !== undefined && str.length > schema.maxLength) {
        throw new Error(
          `String length ${str.length} is greater than maximum ${schema.maxLength}`
        );
      }
      if (schema.pattern) {
        const regex = new RegExp(schema.pattern);
        if (!regex.test(str)) {
          throw new Error(`String does not match pattern: ${schema.pattern}`);
        }
      }
    }

    // Array length validation
    if (schema.type === 'array') {
      const arr = args as unknown[];
      if (schema.minItems !== undefined && arr.length < schema.minItems) {
        throw new Error(
          `Array length ${arr.length} is less than minimum ${schema.minItems}`
        );
      }
      if (schema.maxItems !== undefined && arr.length > schema.maxItems) {
        throw new Error(
          `Array length ${arr.length} is greater than maximum ${schema.maxItems}`
        );
      }
    }
  }

  /**
   * Validates a single value against a schema.
   *
   * @param value - The value to validate
   * @param schema - The schema to validate against
   * @param path - Property path (for error messages)
   * @throws Error if validation fails
   */
  private static validateValue(
    value: unknown,
    schema: JSONSchema,
    path: string
  ): void {
    try {
      ToolExecutor.validateArguments(value, schema);
    } catch (error) {
      throw new Error(
        `Validation error for property '${path}': ${(error as Error).message}`
      );
    }
  }

  /**
   * Executes a tool without schema validation (use with caution).
   *
   * @param handler - The function to execute
   * @param argumentsJson - JSON-encoded arguments
   * @returns Tool execution result as string
   *
   * @example
   * ```typescript
   * const handler = (params: any) => `Got: ${params.value}`;
   * const result = await ToolExecutor.executeUnsafe(
   *   handler,
   *   '{"value": 42}'
   * );
   * ```
   */
  static async executeUnsafe<TParams = Record<string, unknown>>(
    handler: (params: TParams) => string | Promise<string>,
    argumentsJson: string
  ): Promise<string> {
    let args: unknown;
    try {
      args = JSON.parse(argumentsJson);
    } catch (error) {
      throw new Error(
        `Invalid JSON arguments: ${(error as Error).message}`
      );
    }

    try {
      const result = await handler(args as TParams);
      return result;
    } catch (error) {
      throw error;
    }
  }
}
